Bottom: 0c5a0c24885960bf61c5afe1b2ef070808c1f084
Top:    684e21cf3e81b4f5bd740456e0fea103e52d7d1f
Author: Sean Hefty <sean.hefty@intel.com>
Date:   2012-06-04 13:14:42 -0700

rsocket: Handle SHUT_RD/WR shutdown flags

Sridhar Samudrala <sri@us.ibm.com> reported an error (EOPNOTSUPP)
after calling select().

The issue is that rshutdown(SHUT_WR) was called before select().
As part of shutdown, rsockets switches the underlying fd from
nonblocking to blocking to ensure that previously sent data has
completed.  shutdown(SHUT_WR) indicates that the socket should be
kept open for receiving data.

Delay handling the actual shutdown unless SHUT_RDWR is specified,
or the socket is closed.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>


---

diff --git a/src/rsocket.c b/src/rsocket.c
index 8b1e930..a1679f1 100644
--- a/src/rsocket.c
+++ b/src/rsocket.c
@@ -1593,6 +1593,9 @@ int rshutdown(int socket, int how)
 	struct rsocket *rs;
 	int ret = 0;
 
+	if (how != SHUT_RDWR)
+		return 0;
+
 	rs = idm_at(&idm, socket);
 	if (rs->fd_flags & O_NONBLOCK)
 		rs_set_nonblocking(rs, 0);
